### PR TITLE
Adding Variadic Controller Arguments

### PR DESCRIPTION
**Pull request by lukaskleinschmidt**
## This PR …
Currently working on an update for a plugin where I previously used the `Kirby\Toolkit\Controller` to pass in arguments.
Thought this might be something for the core as well, as it does not break anything and allows for some advanced implementations without adding (🤞) any extra complexity. 

Edit:
On second thought this actually fixes some edge cases:

```php
$controller = new Controller(function ($c, ...$b) {
	// Current implementation
	// $c = 'C'
	// $b = ['B']

	// Should be
	// $c = 'C'
	// $b = ['A', 'B']
});

$controller->call(null, [
	'a' => 'A',
	'b' => 'B',
	'c' => 'C'
]);
```

## Ready?
- [x] Unit tests for fixed bug/feature
- [ ] In-code documentation (wherever needed)
- [x] Tests and checks all pass

### For review team
- [x] Add changes to release notes draft in Notion
- [ ] Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)

### CODE REVIEW

File: src/Toolkit/Controller.php [hunk 1]
```diff
@@ -26,11 +26,19 @@ public function __construct(protected Closure $function)
...
+		$args = [];
 
-		return A::map(
-			$info->getParameters(),
-			fn ($parameter) => $data[$parameter->getName()] ?? null
-		);
+		foreach ($info->getParameters() as $param) {
+			$name = $param->getName();
+
+			if ($param->isVariadic()) {
```
Comment (lukasbestle) [inline src/Toolkit/Controller.php]:
```suggestion
			if ($param->isVariadic() === true) {
```

File: tests/Toolkit/ControllerTest.php [hunk 2]
```diff
@@ -29,6 +29,20 @@ public function testArguments()
 		]));
 	}
 
+	/**
+	 * @covers ::arguments
+	 */
+	public function testVariadicArguments()
+	{
+		$controller = new Controller(fn ($c, ...$args) => $c . implode('', $args));
```
Comment (lukasbestle) [inline tests/Toolkit/ControllerTest.php]: To ensure that `$c` still receives its value correctly:

```suggestion
		$controller = new Controller(fn ($c, ...$args) => $c . '/' . implode('', $args));
```

File: tests/Toolkit/ControllerTest.php [hunk 3]
```diff
@@ -29,6 +29,20 @@ public function testArguments()
...
+
+		$this->assertSame('CAB', $controller->call(null, [
```
Comment (lukasbestle) [inline tests/Toolkit/ControllerTest.php]:
```suggestion
		$this->assertSame('C/AB', $controller->call(null, [
```

**Review by lukasbestle — COMMENTED** [review]
I like the idea and also implementation.

Only thought I had is that the name of the variadic argument no longer gets a meaning unlike with normal arguments. But it's probably a non-issue as it changes nothing about the current behavior.

**Comment by lukaskleinschmidt** [comment]
The ues case is probably limited. For reference the origin of the idea:
https://github.com/lukaskleinschmidt/kirby-snippet-controller/issues/1#issuecomment-1567858154

**Review by lukasbestle — DISMISSED** [review]
I like it. But would be good to have a second team opinion on the general concept.

**Comment by distantnative** [comment]
Doesn't seem to impact the existing usage and I'd be fine to have this addition. My head doesn't quite parse how `$args += $data` just adds part of the array, e.g. `$c, ...$args` not also collecting `$c` in `$args`

**Comment by lukaskleinschmidt** [comment]
The `$args += $data` only adds values from `$data` that are not already present in the `$args` array (by key).
It is similar to using `$args = array_merge($args, $data)` in this case as we work with string keys here.

Apparently it is called a union operation.
https://www.php.net/manual/en/language.operators.array.php

**Comment by distantnative** [comment]
I think we are missing a test that checks whether we can use parameters no matter of their order, or am I just missing it? Would be good to ensure that the change doesn't mess with that.

**Comment by lukaskleinschmidt** [comment]
An independend one from the two that test the arguments?

**Comment by distantnative** [comment]
Well, `testArguments` just uses them in the order as they are provided. Would probably be interesting to test `fn( $b, $a)`

**Comment by lukaskleinschmidt** [comment]
I could add another `testArgumentsOrder` test to the PR

**Review by distantnative — APPROVED** [review]

**Review by lukasbestle — APPROVED** [review]

**Comment by bastianallgeier** [comment]
⏸ Merge after 3.9.5 has been released.

**Pull request merged**